content/general/concepts/abstract-factory/abstract-factory.md

```diff
@@ -0,0 +1,381 @@
+
+| Customer Type | Current                     | Saving                |
+| ------------- | --------------------------- | --------------------- |
+| Gold          | Max overdraft limit of 3500 | Interest rate of 5%   |
+| Silver        | Max overdraft limit of 1200 | Interest rate of 3%   |
+| Bronze        | Max overdraft limit of 500  | Interest rate of 1.5% |
+| Builder       | No overdraft                | Interest rate of 1.5% |
+
+To simulate requesting and receiving a customer's credit score, a gateway has been mocked below. When given a customers name, the`CreditAgencyGateway` class should return an appropriate `Customer` object. We can later use this class to see the different paths through our abstract factory.
+
+```java
+public class CreditAgencyGateway {
+  // Depending on which name is searched for, a different customer object is returned
+  public Customer getCustomer(String name) {
+    return switch (name) {
+      case "Harry" -> new Customer("Harry",
+                                   "Potter",
+                                   LocalDate.of(1980, 7, 31),
+                                   "4 Privet Drive, Little Whinging",
+                                   795);
+      case "Ron" -> new Customer("Ron",
+                                 "Weasley",
+                                 LocalDate.of(1980, 3, 1),
+                                 "Shell Cottage, Cornwall",
+                                 379);
+      case "Hermione" -> new Customer("Hermione",
+                                      "Granger",
+                                      LocalDate.of(1979, 9, 19),
+                                      "12 Grimmauld Palace, London",
+                                      843);
+      case "Lucius" -> new Customer("Lucius",
+                                    "Malfoy",
+                                    LocalDate.of(1954, 2, 4),
+                                    "Malfoy Mannor, Wiltshire",
+                                    978);
+      case "Dobby" -> new Customer("Dobby",
+                                   "the house elf",
+                                   LocalDate.of(1972, 6, 28),
+                                   "Malfoy Mannor, Wiltshire",
+                                   256);
+      case "Sirius" -> new Customer("Sirius",
+                                    "Black",
+                                    LocalDate.of(1959, 11, 3),
+                                    "12, Grimmauld Place, London",
+                                    169);
+      default -> throw new IllegalArgumentException("Could not return the credit history for " + name);
+    };
+  }
+}
+```
+
+The `CreditAgencyGateway` uses a switch statement to query a provided `name`. If the name is recognised a new customer is returned, else an exception is thrown.
```

Comment:
```suggestion
The `CreditAgencyGateway` class uses a `switch` statement to query a provided `name`. If the name is recognized a new customer is returned. Otherwise, an exception is thrown.
```